### PR TITLE
 #163756426 User Signup

### PR DESCRIPTION
**Pull request by abayo-luc**
#### What does this PR do?
Add user registration endpoint
#### Description of Task to be completed?
```POST /api/v1/users```
#### How should this be manually tested?
Clone the repo and checkout to ```ft-user-signup-#163756426```
User postman to test the above endpoint. Required attributes are:
```email``` , ```password```, and ```username```
#### Any background context you want to provide?
There are some validation on all fields. All of them are required, and email is checked for validity
#### What are the relevant pivotal tracker stories?
163756426

### CODE REVIEW

File: controllers/users.js [hunk 1]
```diff
@@ -1,33 +1,41 @@
...
-      })
-      .catch(error => {
-        res
-          .status(500)
-          .json({ message: "An error occur please Try again later", error });
       });
+    } catch (error) {
+      if (error.name === "SequelizeUniqueConstraintError") {
+        const { message } = error.errors[0];
+        return res.status(400).json({ message });
```
Comment (espoirMur) [inline controllers/users.js]: @abayo-luc 400 is not the best error code when the user email is already taken...
Please try to search the best error code for this case .....

File: helpers/funcValidators.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export const isEmpty = value => {
```
Comment (espoirMur) [inline helpers/funcValidators.js:1]: @abayo-luc  can we use [Joi](https://www.npmjs.com/package/express-joi-validation) for input validation, please?

File: package.json [hunk 3]
```diff
@@ -4,13 +4,16 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
-    "test": "NODE_ENV=test nyc --reporter=text --reporter=lcov mocha --require babel-polyfill --require @babel/register ./tests --timeout=10000 --exit",
-    "start": "NODE_ENV=development nodemon --exec babel-node index.js"
+    "test": "NODE_ENV=test && yarn db:reset && yarn db:migrate &&  nyc --reporter=text --reporter=lcov mocha --require babel-polyfill --require @babel/register ./tests/**/**.test.js --timeout=10000 --exit",
```
Comment (espoirMur) [inline package.json]: @abayo-luc why `./tests/**/**.test.js` ,  `tests/*`  is enought ...

File: package.json [hunk 4]
```diff
@@ -4,13 +4,16 @@
...
+    "start": "NODE_ENV=development nodemon --exec babel-node index.js",
+    "db:migrate": "node_modules/.bin/sequelize db:migrate",
+    "db:reset": "node_modules/.bin/sequelize db:migrate:undo:all"
   },
   "author": "Andela Simulations Programme",
   "license": "MIT",
```
Comment (espoirMur) [inline package.json:13]: Nice one 👍

File: tests/users/signup.test.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      .request(app)
+      .post("/api/v1/users")
+      .send({ ...userDamie });
+
+    expect(response.status).eql(200);
+    expect(response.body).to.be.an("object");
+    expect(response.body).to.have.property("message");
+    expect(response.body.message).to.be.equals("User registered successfully");
+    expect(response.body.user).to.be.an("object");
+    expect(Object.keys(response.body.user)).to.include.members([
```
Comment (espoirMur) [inline tests/users/signup.test.js:24]: nice one 👍

File: controllers/users.js [hunk 6]
```diff
@@ -1,33 +1,41 @@
...
-    User.findOne({ where: { id: userId } })
-      .then(result => {
-        if (result) {
-          return res.status(200).json({
-            message: "User retrieved successfully",
-            user: result.dataValues
-          });
+  static async signUp(req, res) {
+    const { email, password, username } = req.body;
+    try {
```
Comment (espoirMur) [inline controllers/users.js:11]: Why try and catch and not the promise style??
what is the best way??

File: tests/users/signup.test.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
+import chaiHttp from "chai-http";
+import chai, { expect, should } from "chai";
+import app from "../../index";
+
+const userDamie = {
```
Comment (espoirMur) [inline tests/users/signup.test.js]: @abayo-luc userDamie??? or dummyUser??? Please,  let use meaningful names !

File: tests/users/signup.test.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      .request(app)
+      .post("/api/v1/users")
+      .send({
+        email: "luc.bayo@gmail.com",
+        password: "12345",
+        username: "jean786"
+      });
+    expect(response.status).equal(400);
+    expect(response.body).to.be.an("object");
+    expect(response.body.message).eql("email must be unique");
```
Comment (espoirMur) [inline tests/users/signup.test.js]: is this message meaningful to our users??
Can we use `The email is already taken`? or any other meaningful message?

File: tests/users/signup.test.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    const response = await chai
+      .request(app)
+      .post("/api/v1/users")
+      .send({
+        ...userDamie,
+        email: "jean@andela.com"
+      });
+    expect(response.status).equal(400);
+    expect(response.body).to.be.an("object");
+    expect(response.body.message).eql("username must be unique");
```
Comment (espoirMur) [inline tests/users/signup.test.js]: Something as for email...

File: controllers/users.js (hunk 6 above)
Comment (abayo-luc) [reply controllers/users.js:11]: We are supposed to use async and await

File: tests/users/signup.test.js (hunk 7 above)
Comment (abayo-luc) [reply tests/users/signup.test.js]: I got it!! it was making sense in my head, but I get what you are trying to mean.

File: helpers/funcValidators.js (hunk 2 above)
Comment (abayo-luc) [reply helpers/funcValidators.js:1]: Yeah, we can, but for simple validation like that one, I don't see why I can't stretch my brain and come up with a simple algorithm to do that.

File: .gitignore [hunk 10]
```diff
@@ -44,3 +44,4 @@ package-lock.json
 yarn.*
 yarn-error.*
 .nyc_output
+yarn.lock
```
Comment (niomwungeri-fabrice) [inline .gitignore]: @abayo-luc `yarn.*`  includes `yarn.lock` no need to add it again

File: helpers/funcValidators.js (hunk 2 above)
Comment (niomwungeri-fabrice) [reply helpers/funcValidators.js:1]: I agree with @espoirMur!

Comment (abayo-luc) [reply helpers/funcValidators.js:1]: But you are not giving me reasons why I should delete all of these lines of code and use JOI. Maybe next time for very complicated validation I will consider using JOI, but now I am not getting any reason why I should use JOI.

File: .gitignore (hunk 10 above)
Comment (abayo-luc) [reply .gitignore]: That one is not working properly.

**Review by espoirMur — APPROVED** [review]

**Pull request merged**